File: 1 semester/Task 7-3/Task 7-3/stack.cpp
#include <iostream>
#include "stack.h"

struct StackElement
{
	char value;
	StackElement *next;
};

struct Stack
{
	StackElement *top = nullptr;
};

Stack* createStack()
{
	Stack *stack = new Stack;
	return stack;
}

void push(Stack* stack, char value)
{
	StackElement *stackElement = new StackElement;
	stackElement->value = value;
	stackElement->next = stack->top;
	stack->top = stackElement;
}

char pop(Stack* stack)
{
	if (stack->top == nullptr)
	{
		return ' ';
	}

	int temp = stack->top->value;
	StackElement *stackElement = stack->top;
	stack->top = stack->top->next;
	delete stackElement;
	return temp;
}

bool isEmpty(Stack* stack)
{
	return stack->top == nullptr;
}

char valueElement(Stack* stack)
{
	return stack->top->value;
}

void deleteStack(Stack* stack)
{
	while (!isEmpty)
	{
		pop(stack);
	}

	delete stack;
} 

File: 2 semester/6 homework/Task 6.1/Task 6.1/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Task_6._1
{
    /// <summary>
    /// Представляет коллекцию объектов, действующую по принципу: последним пришел - первым вышел
    /// </summary>
    class Stack
    {
        class StackElement
        {
            public StackElement Next { get; private set; }
            public object Value { get; private set; }

            public StackElement(object value, StackElement next)
            {
                this.Next = next;
                this.Value = value;
            }
        }

        private StackElement top;
        private int size;

        /// <summary>
        ///  Добавление элемента
        /// </summary>
        public void Push(object value)
        {
            StackElement newElement = new StackElement(value, this.top);
            this.top = newElement;
            this.size++;
        }

        /// <summary>
        ///  Чтение верхнего элемента
        /// </summary>
        public object Top()
        {
            return this.top.Value;
        }

        /// <summary>
        /// Удаление всех элементов
        /// </summary>
        public void Clear()
        {
            this.top = null;
            this.size = 0;
        }

        /// <summary>
        ///  Проверка на пустоту
        /// </summary>
        public bool IsEmpty()
        {
            return this.top == null;
        }

        /// <summary>
        ///  Удаление последнего элемента
        /// </summary>
        public object Pop()
        {
            if (IsEmpty())
            {
                return 1;
            }

            object result = this.top.Value;
            this.top = this.top.Next;
            this.size--;
            return result;
        }

        /// <summary>
        ///  Возвращает размер стека
        /// </summary>
        public int Size()
        {
            return this.size;
        }
    }
}
